sys/include/matstat.h

```diff
@@ -0,0 +1,138 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_matstat Matstat - Integer mathematical statistics library
+ * @ingroup     sys
+ * @brief       Library for computing 1-pass statistics
+ *
+ * The Matstat library uses single pass algorithms to compute statistic measures
+ * such as mean and variance over many values. The values can be immediately
+ * discarded after processing, keeping the memory requirement constant
+ * regardless of how many values need to be processed.
+ *
+ * The design goal is to provide basic mathematical statistics operations on
+ * constrained devices with a "good enough" accuracy to be able to provide some
+ * descriptive measures of data. For more accurate measures of statistics, use a
+ * fancier library, or copy the data to a PC.
+ *
+ * It is important to know that using integer operations will result in lower
+ * precision in the computed measures because of truncation.
+ *
+ * @{
+ * @file
+ * @brief       Matstat library declarations
+ *
+ * @author      Joakim Nohlgård <joakim.nohlgard@eistec.se>
+ */
+
+#ifndef MATSTAT_H
+#define MATSTAT_H
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Internal state for computing running statistics
+ */
+typedef struct {
+    int64_t sum;        /**< Sum of values added */
+    uint64_t sum_sq;    /**< Sum of squared values added */
+    uint32_t count;     /**< Number of values added */
+    int32_t min;        /**< Minimum value seen */
+    int32_t max;        /**< Maximum value seen */
+    int32_t offset;     /**< offset value used to improve numerical stability */
+} matstat_state_t;
+
+/**
+ * @brief   Empty state initializer
+ */
+#define MATSTAT_STATE_INIT (const matstat_state_t) { \
+        .min = INT32_MAX, \
+        .max = INT32_MIN, \
+        .count = 0, \
+        .sum = 0, \
+        .sum_sq = 0, \
+        .offset = 0, \
+    }
+
+/**
+ * @brief   Reset state
+ *
+ * @param[in]   state   State struct to clear
+ */
+void matstat_clear(matstat_state_t *state);
+
+/**
+ * @brief   Add a sample to state
+ *
+ * @param[in]   state   State struct to operate on
+ * @param[in]   value   Value to add to the state
+ */
+void matstat_add(matstat_state_t *state, int32_t value);
+
+/**
+ * @brief   Compute the mean value of all samples so far
+ *
+ * The mean may be rounded to the adjacent integer greater than or lower than
+ * the real mean, but the rounding direction is not defined.
+ *
+ * @param[in]   state   State struct to operate on
+ *
+ * @return  arithmetic mean
+ */
+int32_t matstat_mean(const matstat_state_t *state);
+
+/**
+ * @brief   Compute the sample variance of all samples so far
+ *
+ * @param[in]   state   State struct to operate on
+ * @param[in]   mean    mean value (e.g. from @ref matstat_mean)
+ *
+ * @return  sample variance
+ */
+uint64_t matstat_variance(const matstat_state_t *state, int32_t mean);
```

Comment:
Why has the mean to be provided? We are able to calculate it from `state` so why not do it in the function?

Reply:
Saves an expensive 64 bit division if you need both mean and variance

Reply:
A change of algorithm to improve numerical robustness required that the mean be computed for every value added, so I removed the mean function argument